File: leetcode/LetterCombinationsOfAPhoneNumber.java
class Solution {
    private int[] lengthArr = new int[] {0, 0, 3, 3, 3, 3, 3, 4, 3, 4};
    private ArrayList<String> result = new ArrayList<>();
    private int[] array;
    
    public List<String> letterCombinations(String digits) {
        createDigitArray(digits);
        findDigitCombination(0, new StringBuilder());
        
        return result;
    }
    
    private void createDigitArray(String digits) {
        array = new int[digits.length()];
        for(int i = 0; i < array.length; i++) {
            char ch = digits.charAt(i);
            array[i] = ch - '0';
        }
    }
    
    private void findDigitCombination(int k, StringBuilder sb){
        if(k == array.length) {
            if(sb.length() > 0) {
                result.add(sb.toString());
            }
            return;
        }
        
        char[] letters = getDigitToLetters(array[k]);
        for(int i = 0; i < letters.length; i++) {
            sb.append(letters[i]);
            findDigitCombination(k + 1, sb);
            sb.setLength(sb.length() - 1);
        }
    }
    
    private char[] getDigitToLetters(int digit) {
        int length = lengthArr[digit];
        char[] result = new char[length];
        
        int idxAlphabet = 0;
        for(int i = 2; i < digit; i++) {
            idxAlphabet += lengthArr[i];
        }
        
        for(int i = 0; i < length; i++) {
            result[i] = (char) ('a' + (idxAlphabet + i));
        }
        
        return result;
    }
}

File: leetcode/LetterCombinationsofaPhoneNumber.kt
import java.lang.StringBuilder

class LetterCombinations {
    private val result: MutableList<String> = mutableListOf()
    lateinit var digits: String
    private val size by lazy { digits.length }

    fun letterCombinations(digits: String): List<String> {
        this.digits = digits

        recursive(StringBuilder(), 0)
        return result
    }

    private fun recursive(sb: StringBuilder, idx: Int) {
        if (idx == size) {
            if (sb.isNotEmpty())
                result.add(sb.toString())
            return
        }

        val num: Int = digits[idx].digitToInt()

        val start = getStartCharacter(num)
        val cnt = getCharacterCount(num)

        for (i in 0 until cnt) {
            val char = start + i
            val beforeLen = sb.length
            sb.append(char)
            recursive(sb, idx + 1)
            sb.setLength(beforeLen)
        }
    }

    private fun getStartCharacter(num: Int): Char = when(num) {
        2 -> 'a'
        3 -> 'd'
        4 -> 'g'
        5 -> 'j'
        6 -> 'm'
        7 -> 'p'
        8 -> 't'
        9 -> 'w'
        else -> ' '
    }

    private fun getCharacterCount(num: Int) = when(num) {
        7, 9 -> 4
        2, 3, 4, 5, 6, 8 -> 3
        else -> 0
    }
}
